### PR TITLE
Add non regression tests

### PR DESCRIPTION
**Pull request by VincentLanglet**
Closes https://github.com/phpstan/phpstan/issues/5846
Closes https://github.com/phpstan/phpstan/issues/6213
Closes https://github.com/phpstan/phpstan/issues/2567 (in the issue 5846)
Closes https://github.com/phpstan/phpstan/issues/2471

### CODE REVIEW

File: tests/PHPStan/Analyser/data/bug-2471.php [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+		return 'foo';
+	}
+}
+
+/** @var A[] */
+$y = doFoo();
+
+$x = array_fill_keys($y, null);
+
+assertType('array<string, null>', $x);
```
Comment (staabm) [inline tests/PHPStan/Analyser/data/bug-2471.php]: there should not be any code at the top level of a test. please move it into a method (same in other test files)

Comment (VincentLanglet) [reply tests/PHPStan/Analyser/data/bug-2471.php]: Sure, but can you explain me why ?

Comment (staabm) [reply tests/PHPStan/Analyser/data/bug-2471.php]: IIRC code on the top level would be executed when running the test suite. but we just want phpstan to analyse it

**Comment by VincentLanglet** [comment]
Regressions tests added @ondrejmirtes  =)

File: tests/PHPStan/Rules/Comparison/data/bug-6213.php [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+class HelloWorld
+{
+	public function sayHello(): void {
+		$document = new DOMDocument('1.0', 'utf-8');
+		$element = $document->createElement('div', 'content');
+
+		// Incorrect! This is DOMNode|null not DOMElement|null
+		// It's also possible to contain a DOMText node, which is not an instance
+		// of DOMElement, but an instance of DOMNode!
+		if ($element->firstChild instanceof DOMText) {
```
Comment (ondrejmirtes) [inline tests/PHPStan/Rules/Comparison/data/bug-6213.php]: This test would never fail. The error message is: `Instanceof between DOMElement|null and DOMText will always evaluate to false.`

This test needs to be in ImpossibleInstanceOfRuleTest. Please verify that locally by reverting the stubs update commit. Thanks.

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]

File: tests/PHPStan/Rules/Comparison/data/bug-6213.php (hunk 2 above)
Comment (VincentLanglet) [reply tests/PHPStan/Rules/Comparison/data/bug-6213.php]: I updated the test

It fails without the stub update
image

**Comment by ondrejmirtes** [comment]
Thank you very much!

**Pull request merged**